### ISSUE TITLE
Help with usung the scripts in various emulators

### ISSUE THREAD

**Issue by Peelio** [KeeyanGhoreshi/PokemonPlatinumSingleSequence#3]
G'day
I'm not sure if this is an appropriate place to ask this question, but I'm not very familiar with computer programming or GitHub, so I thought maybe this would be a good place to gain some clarity. If this is not a good place to do this, I'm sorry, and any nudges in the right direction would be wicked.

I love the idea of being able to boot up an emulator, load up the inputs, and just have a solved Pokémon game running in the background. But after a small amount of fiddling around, I have been unable to get the scripts to run in my emulator (DeSmuME). Is this simply because the scripts were written to work with your stripped-down emulator, or am I just missing something that is required?

I loved the video, and it blows my mind every time I think about it and how you've found a set of inputs that completes every single possible version of the game. Just crazy stuff.

Sorry again if this is a waste of time/space on this board.

**Comment by AdmiralCurtiss** [comment]
It's a Bizhawk lua script. You want to download https://tasvideos.org/Bizhawk, launch the game in that, then go to to Tools -> Lua Console, Script -> Open Script -> `startAll.lua`.

**Comment by ZacharyGiguere** [comment]
I'm using Bizhawk 2.9.1 and I have issues with running the script. The save and savestate commands don't seems to work properly on my computer, since I can't find any trace of any saves done through 1.lua with the 'save' command. The savestate.saveslot(id) done before executing the script seems to work (I get Pokemon Platinum Version (US).MelonDS.QuickSave1.State updated each time I start the script) but it keeps crashing when trying to do the save after the script savestate.saveslot(id + 1). In the lua console it says that the slotNum parameter must be between 1 and 10. Am I the only one who this doesn't work?

**Comment by KeeyanGhoreshi** [comment]
that's strange, the id+1 can only ever be 2 or 4, so no idea why that's failing for you.  

To make use of the save function, you have to update the save directory in movement.lua:

https://github.com/KeeyanGhoreshi/PokemonPlatinumSingleSequence/blob/457fc21ded8791646edf976184bdbe51893b36da/commands/movement.lua#L659

Change the path to wherever you want the autosaves to save the files.  Only change this part `'../../../NDS/State/Platinum/autosaves/' and leave the rest alone.

If you can get the autosaves working, the savestate/saveslots become irrelevant.

**Comment by ZacharyGiguere** [comment]
Ok so I already found about the path, instead of changing it I just put the script somewhere the path would be valide (and created the missing folders). So the path NDS/State/Platinum/autosaves/ exist, maybe my starting location for the path is wrong and because of that it can't find the NDS folder...

Also, if autosaves work I wouldn't need savestate/saveslots, but my main issue is the crash. Obviously, the crash don't occure because of ID being 2 or 4. Infact, to be sure it was working the first time and not just a default mechanic, I changed the value of ID to 2 instead of 1 for the first savestate of 1.lua and it correctly create Pokemon Platinum Version (US).MelonDS.QuickSave2.State. The crash occurs more probably because of ID getting outside of the range or maybe being not even a number anymore. I am not familiar to lua, but if variable can be global than the value of ID get changed here 
https://github.com/KeeyanGhoreshi/PokemonPlatinumSingleSequence/blob/457fc21ded8791646edf976184bdbe51893b36da/commands/movement.lua#L660
I'm unsure why ID is asigned the value of client.xpos() in the first place, but if it also affect the value of ID in 1.lua than the number is clearly out of range

Last, unless there are autosaves made somewhere I didn't saw on my PC, it looks like the save commands don't work and don't output error, which is strange and it is probably something on my side, since everyone else seems to have it working. The only odd thing I noticed in the name for the autosave is the `id .. '_' .. '_' .. fileName` where there is 2 _ without anything between, but it shouldn't affect anything.

**Comment by KeeyanGhoreshi** [comment]
Nice, variables in lua are global unless declared local so you're right, the save function is overwriting the id.  What's interesting is that on my emulator, this issue doesn't cause a crash.  It doesn't even produce an error! Logging the id before it tries to save indicates that it is obviously being overwritten, but I don't see any log and everything continues running as expected.  

Maybe my version of bizhawk is different and it doesn't throw an error when the saveslot function doesn't work.  Either way, this is a pretty simple fix.  I was originally going to make the id local, but the saveslots have no use anymore to anybody so I'm just going to remove them.  

As an aside, the reason id is set to xpos is to auto-generate a unique id for each emulator I had running on my pc, since I'd run them side by side.  There also used to be an additional piece of information saved in each state that I removed, but just forgot to remove the second underscore.  

As for the autosaves, I'd recommend just specifying the path you want directly.  Also note that the autosave function saves into a directory depending on which script is running.  You will want your directory to look like this:

![image](https://github.com/user-attachments/assets/7d38652f-aef2-47b5-a82c-46c87e36ffd7)

**Comment by ZacharyGiguere** [comment]
The explanation for the xpos was my guess lol. Creating the script folders fix the issue! And since the new version completely remove the saveslot it should no longer crash! Thanks a lot for the help, but most importantly for realising this project!!

**Comment by KeeyanGhoreshi** [comment]
Yep, you'll have to comment out everything before the autosave point so that it starts up from that position.  I usually just put the second section into another array called `b` or `c` and pass it to `movement.executeOrders` instead so that I don't have to remember to uncomment a bunch of stuff.

**Comment by hitgub043** [comment]
I'm not sure how to get the script to run. went into Bizhawk's Lua console and loaded the startAll script, while the game was running, and nothing is happening. The game just remains on the main menu, cycling through the opening movie. Does the script not press Start to begin the game? 
The Output window in the console is also empty. Should there be something here when the script is running?

I'm also wondering about the timing of the script. Do I need to have the script run immediately as the game powers on? If so, how?

**Comment by KeeyanGhoreshi** [comment]
The script assumes that you are playing a fresh game of platinum with no current save. If you have a save file already, that could cause this issue. 

The output window should remain empty. The beginning of the script expects to start essentially any time at the opening movie, but you can start the script from the very beginning by pausing the emulation, rebooting the core (an option under the “emulation” tab) and then running the script. 

It’s also possible the script isn’t running at all. I would recommend making sure “View->Display Input” in Bizhawk is turned on, so that you can see what inputs the script is making. It will also tell you if it’s running. 

If it is actually sending inputs, then the issue might be related to the version of your emulator.

**Comment by hitgub043** [comment]
Thanks! With input display on, I can tell that the script is running. (If only the emulator itself had an indicator for that!)

However, I seem to have a problem with the touch control now. I can see in the input display that the "T" command is being entered, but the game doesn't progress past the "touch the button on the Pokeball" part of the intro. I'm guessing that the command is either being ignored, or the game doesn't think the touch is happening at the correct location.

Is there a specific way to make this work? The readme mentions that it could be a problem, but I'm not sure what the solution is.
If it's relevant, my input display shows the touchscreen coordinates, based on the mouse cursor position. These coordinates are updated even if the mouse is moved OUTSIDE the game window, so there's nothing I can do to prevent them from changing. Not sure if this is related to the command not working correctly.

EDIT: If I open the Virtual Gamepad (the on-screen controls), the touch coordinates no longer update based on mouse position. HOWEVER, when the script performs the Touch command, the coordinates are set to 255,191, which is the far-right corner of the screen. Does this indicate that the script touching the wrong point?

**Comment by KeeyanGhoreshi** [comment]
The emulators analog controls are a little finicky. I’ll edit this comment when I have a chance to look at the Bizhawk menus so I can provide better advice. But in the meantime, you can try messing with the analog controls. What’s most likely happening is that the emulator is ignoring the coordinates of the touch command and instead using the current coordinates based on your mouse position. If your mouse is off the emulator screen, it will just use the last coordinate prior to your mouse leaving the window (which is why it reads as the bottom right of the screen). 

I remember fixing this problem myself by turning on/off some of the analog control settings but I can’t remember the specifics.

**Comment by KeeyanGhoreshi** [comment]
Checking in the emulator, clearing the input controls for analog + touch allow it to work. In Bizhawk that’s config -> controllers -> analog controls, and then unbind the touch X and touch Y, then for good measure under normal controls, unbind/clear the touch control (that might not be necessary). After that, it should work as intended.

**Comment by hitgub043** [comment]
Thanks! I just manually pressed the pokeball this time, but I'll try editing out the control mappings next time I run the emulation. :)
